packages/x-data-grid-pro/src/hooks/features/treeData/treeDataReorderExecutor.ts

```diff
@@ -0,0 +1,741 @@
+    apiRef.current.updateRows([updatedRow]);
+    apiRef.current.publishEvent('rowsSet');
+  }
+}
+
+/**
+ * Handles dropping any node (leaf or group) "over" a leaf node.
+ * This converts the target leaf into a parent group and makes the dragged node its child.
+ */
+class DropOnLeafOperation extends BaseReorderOperation {
+  readonly operationType = 'drop-on-leaf';
+
+  detectOperation(ctx: ReorderExecutionContext): ReorderOperation | null {
+    const {
+      sourceRowId,
+      dropPosition,
+      placeholderIndex,
+      sortedFilteredRowIds,
+      apiRef,
+      setTreeDataPath,
+    } = ctx;
+
+    // Only applies to "over" drop position
+    if (dropPosition !== 'over') {
+      return null;
+    }
+
+    const sourceNode = gridRowNodeSelector(apiRef, sourceRowId);
+    if (!sourceNode || sourceNode.type === 'footer') {
+      return null;
+    }
+
+    // Check if setTreeDataPath is provided
+    if (!setTreeDataPath) {
+      if (process.env.NODE_ENV !== 'production') {
+        warnOnce(
+          'MUI X: Drop on leaf reordering requires `setTreeDataPath()` prop to update row data paths. ' +
+            'Please provide a `setTreeDataPath()` function to enable this feature.',
+          'warning',
+        );
+      }
+      return null;
+    }
+
+    // Find target node
+    let targetIndex = placeholderIndex;
+    if (targetIndex >= sortedFilteredRowIds.length && sortedFilteredRowIds.length > 0) {
+      targetIndex = sortedFilteredRowIds.length - 1;
+    }
+
+    if (targetIndex < 0) {
+      return null;
+    }
+
+    const targetNode = gridRowNodeSelector(apiRef, sortedFilteredRowIds[targetIndex]);
+    if (!targetNode || targetNode.type !== 'leaf') {
+      return null;
+    }
+
+    // Target leaf will become a parent, so the actual target index is 0 (first child)
+    const actualTargetIndex = 0;
+
+    return {
+      sourceNode,
+      targetNode,
+      actualTargetIndex,
+      isLastChild: false,
+      operationType: 'drop-on-leaf',
+    };
+  }
+
+  async executeOperation(operation: ReorderOperation, ctx: ReorderExecutionContext): Promise<void> {
+    const { sourceNode, targetNode } = operation;
+    const { apiRef, setTreeDataPath, processRowUpdate, onProcessRowUpdateError } = ctx;
+
+    const rowTree = gridRowTreeSelector(apiRef);
+    const dataRowIdToModelLookup = gridRowsLookupSelector(apiRef);
+
+    // Build target path for the new structure
+    const targetPath = buildTreeDataPath(targetNode, rowTree);
+    let rowsToUpdate: GridValidRowModel[] = [];
+
+    // Handle source node path updates
+    if (sourceNode.type === 'leaf') {
+      // Simple leaf move
+      const leafKey = sourceNode.groupingKey;
+      const newPath = [...targetPath, String(leafKey)];
+
+      const originalRow = dataRowIdToModelLookup[sourceNode.id];
+      let updatedRow = setTreeDataPath!(newPath, originalRow);
+
+      // Process row update if needed
+      if (processRowUpdate) {
+        try {
```

Comment:
I see that the same (or very similar) try...catch block being repeated multiple times, could you deduplicate it?
Also, why isn't `BatchRowUpdater` used for all updates? Even when we know that only one row will be updated, it might still make sense if we let the updater handle the process. WDYT?